Transfer counters and metrics

FloatingCrowbar opened the pull request:
### Changelog entry <!-- a user-readable short description of the changes that goes to CHANGELOG.md and Release Notes -->

Transfer metrics and stats in describe for better observability

### Changelog category <!-- remove all except one -->

* New feature

### Description for reviewers <!-- (optional) description for those who read this PR -->

FloatingCrowbar committed:
Transfer counters - WIP

alexnick88 commented on ydb/core/protos/replication.proto:
```diff
@@ -115,6 +116,7 @@ message TReplicationConfig {
         optional TTarget Target = 1;
         optional TBatchingSettings Batching = 2;
         optional string RunAsUser = 3;
+        optional uint64 CountersLevel = 4;
```
MetricsLevel у нас называется опция

alexnick88 commented on ydb/core/tx/replication/service/topic_reader.cpp:
```diff
@@ -58,20 +62,54 @@ class TRemoteTopicReader: public TActor<TRemoteTopicReader> {
+            }
+            if (compressed) {
+                if (!descompressRequired) {
+                    SendOperationChange(EWorkerOperation::DECOMPRESS);
```
На каждый чих отсылать эвент зачем?  Стоит насчитывать состояние и отправлять инфу в контроллер раз в сколько-то времени, а не по факту смены состояния.

FloatingCrowbar commented on ydb/core/tx/replication/service/topic_reader.cpp:
```diff
@@ -58,20 +62,54 @@ class TRemoteTopicReader: public TActor<TRemoteTopicReader> {
+
+        if (Settings.ReportStats_) {
+            Y_ABORT_UNLESS(!Requests.empty());
+            auto request = Requests.front();
```
По нынешней логике, кажется, не должно быть такого. Но технически ничего не мешает в этот актор отправить несколько EvPoll подряд, и он даже нормально их отработает. Ломать это введением статистики не хотелось бы.

FloatingCrowbar committed:
Functional, w/o tests

CyberROFL commented on ydb/core/protos/counters_replication.proto:
Лишний дифф.

CyberROFL commented on ydb/core/transfer/transfer_writer.h:
```diff
@@ -5,11 +5,21 @@
-namespace NKikimr::NReplication::NTransfer {
+namespace NKikimr::NReplication {
+
+struct TTransferWriteStats {
```
Почему это не в `NTransfer`?

CyberROFL commented on ydb/core/tx/replication/controller/controller.cpp:
```diff
@@ -446,7 +446,9 @@ void TController::Handle(TEvService::TEvStatus::TPtr& ev, const TActorContext& c
-
+    if (ev->Get()->Record.HasStats()) {
+        Cerr << "TEvService::TEvWorkerStatus: " << ev->Get()->Record.DebugString() << Endl;
+    }
```
Убери дебажную печать.

CyberROFL commented on ydb/core/tx/replication/controller/controller.cpp:
```diff
@@ -506,6 +510,19 @@ void TController::UpdateLag(const TWorkerId& id, TDuration lag) {
+    if (!target) {
+        return;
+    }
+    target->UpdateStats(id.WorkerId(), stats, AppData()->Counters);
```
```suggestion

    target->UpdateStats(id.WorkerId(), stats, AppData()->Counters);
```

CyberROFL commented on ydb/public/api/protos/draft/ydb_replication.proto:
```diff
@@ -122,6 +133,72 @@ message DescribeTransferResult {
+            // writing data to table
+            STATE_WRITE = 4;
+        }
+        message WorkerStats {
```
```suggestion

        message WorkerStats {
```

CyberROFL commented on ydb/public/api/protos/draft/ydb_replication.proto:
```diff
@@ -122,6 +133,72 @@ message DescribeTransferResult {
+            // cpu usage time on processing, milliseconds
+            MultipleWindowsStat processing_cpu_time_ms = 14;
+        }
+        // detailed stats of every worker, only included if DescribeTransferRequest.include_stats flag = true in request
```
```suggestion

        // detailed stats of every worker, only included if DescribeTransferRequest.include_stats flag = true in request
```

CyberROFL commented on ydb/core/tx/replication/service/worker.cpp:
```diff
@@ -141,6 +159,7 @@ class TWorker: public TActorBootstrapped<TWorker> {
         ui32 GetCreateAttempt() const {
             return CreateAttempt;
         }
+
```
Лишний дифф.

CyberROFL commented on ydb/core/tx/replication/service/topic_reader.h:
```diff
@@ -2,8 +2,18 @@
+    ui64 TotalDecompressTimeMs = 0;
+    ui64 ReadCpuMs = 0;
+    ui64 DecompressCpuMs = 0;
+};
```
Я бы это отдельно положил, потому что сейчас это имеет смысл только для транфера.

FloatingCrowbar replied:
Перенес в другой header

CyberROFL commented on ydb/core/tx/replication/controller/event_util.cpp:
```diff
@@ -43,9 +45,13 @@ THolder<TEvService::TEvRunWorker> MakeRunWorkerEv(
+    if (metricsLevel) {
+        command->SetMetricsLevel(metricsLevel);
+    }
+    auto& readerSettings = *command->MutableRemoteTopicReader();
```
```suggestion

    auto& readerSettings = *command->MutableRemoteTopicReader();
```

CyberROFL commented on ydb/core/protos/replication.proto:
```diff
@@ -308,6 +330,41 @@ message TEvStopWorker {
+    message TWorkerStatItem {
+        optional uint64 Key = 1;
+        optional int64 Value = 2;
+    }
```
```suggestion
    }

```

CyberROFL commented on ydb/core/tx/replication/controller/replication.h:
```diff
@@ -4,6 +4,7 @@
 
 #include <ydb/core/base/defs.h>
 #include <ydb/core/scheme/scheme_pathid.h>
+#include <ydb/core/protos/replication.pb.h>
```
Не надо сюда тащить зависимость на протобуф.

CyberROFL commented on ydb/core/tx/replication/controller/tx_describe_replication.cpp:
```diff
@@ -101,6 +101,7 @@ class TTargetDescriber: public TActorBootstrapped<TTargetDescriber> {
 
 }; // TTargetDescriber
 
+
```
Лишний дифф.

CyberROFL commented on ydb/core/tx/replication/service/service.h:
```diff
@@ -3,8 +3,8 @@
 #include <ydb/core/base/row_version.h>
-#include <ydb/core/protos/replication.pb.h>
 #include <ydb/core/tx/replication/common/sensitive_event_pb.h>
+#include <ydb/core/protos/replication.pb.h>
```
Зачем?

CyberROFL commented on ydb/core/tx/replication/service/service.h:
```diff
@@ -75,10 +75,26 @@ struct TEvService {
 
-    struct TEvWorkerDataEnd: public TEventPB<TEvWorkerDataEnd, NKikimrReplication::TEvWorkerDataEnd, EvWorkerDataEnd> {
-        TEvWorkerDataEnd() = default;
+        struct TEvWorkerDataEnd: public TEventPB<TEvWorkerDataEnd, NKikimrReplication::TEvWorkerDataEnd, EvWorkerDataEnd> {
```
Зачем?

FloatingCrowbar replied:
Случайно :) Почистил

CyberROFL commented on ydb/core/tx/replication/service/topic_reader.cpp:
```diff
@@ -1,9 +1,10 @@
 #include "logging.h"
-#include "topic_reader.h"
 #include "worker.h"
+#include "topic_reader.h"
```
Зачем?

FloatingCrowbar replied:
Иначе не собирается - incomplete type 'NKikimr::NReplication::TTransferReadStats'

CyberROFL reviewed (changes requested):
Я уже под конец устал писать однотипные комментарии.

CyberROFL commented:
Призовите меня явно, когда будет всё готово. Сейчас это смотреть невозможно.

FloatingCrowbar commented on ydb/core/protos/replication.proto:
```diff
@@ -308,6 +317,44 @@ message TEvStopWorker {
+        optional int64 Value = 2;
+    }
+    optional EWorkOperation CurrentOperation = 1;
+    repeated TWorkerStatItem Values = 3;
```
Для гибкости. Видов репликации много, в будущем кто-то другой может захотеть воспользоваться механизмом - так проще будет добавить просто новый ключ.

FloatingCrowbar committed:
Update ydb/core/kqp/provider/yql_kikimr_gateway.h

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

FloatingCrowbar committed:
Update ydb/core/tx/replication/controller/controller.cpp

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

FloatingCrowbar committed:
Update ydb/public/api/protos/draft/ydb_replication.proto

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

FloatingCrowbar committed:
Style improvements

FloatingCrowbar committed:
Update ydb/public/api/protos/draft/ydb_replication.proto

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

FloatingCrowbar committed:
Style

FloatingCrowbar committed:
Style

FloatingCrowbar committed:
Update ydb/core/protos/replication.proto

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

FloatingCrowbar committed:
Update ydb/core/protos/replication.proto

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

FloatingCrowbar committed:
Update ydb/core/protos/replication.proto

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

FloatingCrowbar committed:
Update ydb/core/protos/replication.proto

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

FloatingCrowbar committed:
Update ydb/core/tx/replication/controller/event_util.cpp

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

FloatingCrowbar committed:
Update ydb/core/tx/replication/controller/event_util.cpp

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

FloatingCrowbar committed:
Refactor

FloatingCrowbar committed:
Finalize work and add tests

FloatingCrowbar committed:
Fix tests

FloatingCrowbar committed:
Style

nshestakov commented on ydb/core/kqp/ut/scheme/kqp_scheme_ut.cpp:
```diff
@@ -10154,6 +10154,86 @@ Y_UNIT_TEST_SUITE(KqpScheme) {
+        auto repl = TReplicationClient(kikimr.GetDriver(), TCommonClientSettings().Database("/Root"));
+
+        {
+            auto query = R"(
```
а какой смысл этой проверки в этом тесте?

nshestakov commented on ydb/core/protos/replication.proto:
```diff
@@ -117,8 +119,20 @@ message TReplicationConfig {
     }
 
+    message TMetricsConfig {
+        enum EMetricsLevel {
```
почему enum не общий для всех объектов?

nshestakov commented on ydb/core/transfer/ut/common/utils.h:
```diff
@@ -233,6 +236,114 @@ inline TMessage _withAttributes(std::map<std::string, std::string>&& attributes)
+                result = iter->second;
+            }
+        }
+        else {
```
лишний перевод строки

nshestakov commented on ydb/core/tx/replication/controller/target_transfer.cpp:
```diff
@@ -77,4 +86,214 @@ const TString& TTargetTransfer::TTransferConfig::GetDirectoryPath() const {
+    if (newStats.HasStartTime()) {
+        workerStats.StartTime = TInstant::Seconds(newStats.GetStartTime().seconds());
+    }
+    bool hasWorkerWithoutStartTime = false;
```
единственное использование с отрицанием. лучше переименовать в allWorkersWithStartTime и инвертировать логику

nshestakov commented on ydb/core/tx/replication/controller/target_transfer.cpp:
```diff
@@ -77,4 +86,214 @@ const TString& TTargetTransfer::TTransferConfig::GetDirectoryPath() const {
+            }
+        }
+    }
+    if ((!hasWorkerWithoutStartTime && Stats->LastWorkerStartTime == TInstant::Zero()) || Stats->LastWorkerStartTime < workerStats.StartTime) {
```
я не понимаю этот if.

почему нельзя написать?
```
Stats->LastWorkerStartTime = std::max(Stats->LastWorkerStartTime, workerStats.StartTime)
```

FloatingCrowbar commented on ydb/core/kqp/ut/scheme/kqp_scheme_ut.cpp:
```diff
@@ -10154,6 +10154,86 @@ Y_UNIT_TEST_SUITE(KqpScheme) {
+        {
+            auto query = Sprintf(R"(
+                --!syntax_v1
+                CREATE TRANSFER `/Root/transfer`
```
Можно конечно

nshestakov commented on ydb/core/tx/replication/controller/target_table.h:
```diff
@@ -10,6 +10,10 @@ class TTargetTableBase: public TTargetWithStream {
+    void WorkerStatusChanged(ui64 workerId, ui64 status, NMonitoring::TDynamicCounterPtr counters) override;
+    void UpdateStats(ui64, const NKikimrReplication::TWorkerStats&, NMonitoring::TDynamicCounterPtr) override;
+
+    const TReplication::ITargetStats* GetStats() const override;
```
Обсудил с Ильназом. Он согласен, что метрики, которые общие для трансфера и асинхронной репликации, должны быть реализованы и для асинхронной репликации тоже.

FloatingCrowbar replied:
Переписал

CyberROFL commented on ydb/core/grpc_services/rpc_replication.cpp:
```diff
@@ -168,8 +168,8 @@ class TDescribeReplicationRPC: public TRpcSchemeRequestActor<TDescribeReplicatio
 
-    static void BuildRequest(const Replication::DescribeTransferRequest*, NKikimrReplication::TEvDescribeReplication&) {
-        // nop
+    static void BuildRequest(const Replication::DescribeTransferRequest* from, NKikimrReplication::TEvDescribeReplication& to) {
```
Эта функция идентична предыдущей. Можно сделать одну шаблонную функцию.

CyberROFL commented on ydb/core/tx/replication/controller/controller.cpp:
```diff
@@ -506,6 +520,23 @@ void TController::UpdateLag(const TWorkerId& id, TDuration lag) {
+    if (!target) {
+        return;
+    }
+    target->WorkerStatusChanged(id.WorkerId(), status, AppData()->Counters);
```
```suggestion

    target->WorkerStatusChanged(id.WorkerId(), status, AppData()->Counters);
```

CyberROFL commented on ydb/core/tx/replication/service/service.h:
```diff
@@ -75,6 +76,21 @@ struct TEvService {
+                auto* val = stats->AddValues();
+                val->SetKey(k);
+                val->SetValue(v);
+            }
```
```suggestion
            auto& stats = *Record.MutableStats();
            if (startTime) {
                stats.MutableStartTime()->set_seconds(startTime.Seconds());
            }
            for (auto [k, v] : statsValues) {
                auto& val = *stats.AddValues();
                val.SetKey(k);
                val.SetValue(v);
            }
```

CyberROFL commented on ydb/core/tx/replication/service/worker.cpp:
```diff
@@ -277,6 +303,23 @@ class TWorker: public TActorBootstrapped<TWorker> {
+    void Handle(TEvWorker::TEvStatus::TPtr& ev) {
+        LOG_D("Handle " << ev->Get()->ToString());
+        if (!ev->Get()->DetailedStats) {
+            LOG_W("Unexpected TEvWorker::TEvStatus with no stats, ignored. Sender# " << ev->Sender);
```
```suggestion
            LOG_W("Unexpected TEvWorker::TEvStatus with no stats, ignored"
                << ": sender# " << ev->Sender);
```

CyberROFL commented on ydb/core/tx/replication/service/worker.cpp:
```diff
@@ -277,6 +303,23 @@ class TWorker: public TActorBootstrapped<TWorker> {
+
+    TEvWorker::TEvStatus* MakeEvStatusFromReaderStats(std::unique_ptr<TWorkerDetailedStats>&& stats) const {
+        Y_ENSURE(stats->ReaderStats);
+        auto* ev = TEvWorker::TEvStatus::FromOperation(EWorkerOperation::NONE);
```
Лучше возвращать умный указатель (`THolder` или `std::unique_ptr`).

CyberROFL commented on ydb/public/sdk/cpp/include/ydb-cpp-sdk/client/draft/ydb_replication.h:
```diff
@@ -191,6 +191,18 @@ struct TBatchingSettings {
+        Process = 3,
+        Write = 4,
+    };
+    EWorkOperation Operation;
```
```suggestion

    EWorkOperation Operation;
```

CyberROFL commented on ydb/core/tx/schemeshard/schemeshard__operation_create_replication.cpp:
```diff
@@ -162,6 +162,22 @@ class TConfigureParts: public TSubOperationState {
+                }
+                if (auto it = attrs.find("monitoring_project_id"); it != attrs.end()) {
+                    location->SetMonitoringProjectId(it->second);
+                }
```
```suggestion
                auto& location = *ev->Record.MutableLocation();
                location.SetPath(TPath::Init(pathId, context.SS).PathString());

                const auto& attrs = context.SS->PathsById.at(context.SS->RootPathId())->UserAttrs->Attrs;
                if (auto it = attrs.find("cloud_id"); it != attrs.end()) {
                    location.SetYcCloudId(it->second);
                }
                if (auto it = attrs.find("folder_id"); it != attrs.end()) {
                    location.SetYcFolderId(it->second);
                }
                if (auto it = attrs.find("database_id"); it != attrs.end()) {
                    location.SetYdbDatabaseId(it->second);
                }
                if (auto it = attrs.find("monitoring_project_id"); it != attrs.end()) {
                    location.SetMonitoringProjectId(it->second);
                }
```

CyberROFL commented on ydb/core/tx/schemeshard/schemeshard__operation_alter_replication.cpp:
```diff
@@ -117,6 +117,21 @@ class TConfigureParts: public TSubOperationState {
+                }
+                if (auto it = attrs.find("monitoring_project_id"); it != attrs.end()) {
+                    location->SetMonitoringProjectId(it->second);
+                }
```
```suggestion
                auto& location = *ev->Record.MutableLocation();
                location.SetPath(TPath::Init(pathId, context.SS).PathString());

                const auto& attrs = context.SS->PathsById.at(context.SS->RootPathId())->UserAttrs->Attrs;
                if (auto it = attrs.find("cloud_id"); it != attrs.end()) {
                    location.SetYcCloudId(it->second);
                }
                if (auto it = attrs.find("folder_id"); it != attrs.end()) {
                    location.SetYcFolderId(it->second);
                }
                if (auto it = attrs.find("database_id"); it != attrs.end()) {
                    location.SetYdbDatabaseId(it->second);
                }
                if (auto it = attrs.find("monitoring_project_id"); it != attrs.end()) {
                    location.SetMonitoringProjectId(it->second);
                }
```

CyberROFL reviewed (changes requested):
Не дочитал, слишком много...

FloatingCrowbar committed:
Update ydb/core/tx/replication/controller/controller_impl.h

style

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

FloatingCrowbar committed:
Update ydb/core/tx/replication/service/topic_reader.cpp

Log formatting

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

FloatingCrowbar committed:
Update ydb/core/tx/replication/service/topic_reader.cpp

Style

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

FloatingCrowbar committed:
Update ydb/core/tx/replication/controller/replication.h

Naming

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

FloatingCrowbar committed:
Update ydb/core/tx/replication/controller/controller.cpp

style

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

FloatingCrowbar committed:
Update ydb/core/tx/replication/controller/tx_describe_replication.cpp

Naming

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

FloatingCrowbar committed:
Update ydb/core/tx/replication/controller/tx_describe_replication.cpp

Style

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

FloatingCrowbar committed:
Update ydb/core/tx/replication/service/service.h

style (whitespaces)

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

FloatingCrowbar committed:
Update ydb/core/tx/replication/service/service.h

Style

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

FloatingCrowbar committed:
Update ydb/core/tx/replication/controller/target_transfer.h

Style

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

FloatingCrowbar committed:
Update ydb/core/tx/replication/controller/target_transfer.h

Naming

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

FloatingCrowbar committed:
Apply suggestions from code review

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

pixcc committed:
Add smart mode to CMS (#33297)

FloatingCrowbar committed:
Refactor targets and few improvements

FloatingCrowbar committed:
Reset default metrics level

nshestakov commented on ydb/core/tx/replication/controller/target_transfer.h:
```diff
@@ -2,37 +2,50 @@
+
 private:
     TActorId StreamConsumerRemover;
+    ui64 MetricsLevel = 0;
```
ниже используется перечисление. почему здесь число?

nshestakov commented on ydb/core/tx/replication/service/service.cpp:
```diff
@@ -642,6 +842,30 @@ class TReplicationService: public TActorBootstrapped<TReplicationService> {
+                    sessionIter->second.SendWorkersStats(this);
+                }
+            }
+            LastStatsUpdate = Now();
```
лучше хранить когда следующий раз надо выполнить, а не время последнего. тогда при большом кол-ве событий + StatsWakeupInterval быдет выполняться только один раз

FloatingCrowbar committed:
Minor improvements after review

FloatingCrowbar committed:
Slight refactor

ijon committed:
schemeshard: add batch processing for ttl responses (#33867)

Changes:
- Add size- and time- based batching for `TEvConditionalEraseRowsResponse`. `TTxScheduleConditionalErase` now processes batches instead of individual responses.
- Process `TEvConditionalEraseRowsResponse::PARTIAL` response early, without entering a local transaction.
- Also move TSchemeShardConfig from `config.proto` to separate `schemeshard_config.proto`.

Configuration:
- `feature_flags.enable_conditional_erase_response_batching` — enables batching
- `schemeshard_config.cond_erase_response_batch_size` — batch size limit
	- default: 100
	- zero disables batching
- `schemeshard_config.cond_erase_response_batch_max_time_ms` — max pending batch time
	- default: 100 ms
	- value clamped between 1-1000 ms

CyberROFL commented on ydb/core/tx/replication/controller/target_with_stream.h:
```diff
@@ -2,24 +2,97 @@
+
+
+public:
+    TTargetWithStreamStats(TInstant startTime);
```
```suggestion
    explicit TTargetWithStreamStats(TInstant startTime);
```

CyberROFL reviewed (changes requested):
Я опять досмотрел не всё... Слишком много.

FloatingCrowbar committed:
Apply suggestions from code review

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

FloatingCrowbar committed:
Review improvements

CyberROFL commented on ydb/core/tx/replication/service/service.cpp:
```diff
@@ -99,11 +203,23 @@ class TSessionInfo {
+                            TMetricsConfig::EMetricsLevel metricsLevel)
+    {
+        auto res = Workers.emplace(id, TWorkerInfo{ops->Register(actor, TMailboxType::HTSwap, poolId),
+                                                   replicationLocation, metricsLevel, id.WorkerId(), AppData()->Counters});
```
Выравнивание.

CyberROFL commented on ydb/core/tx/replication/service/service.cpp:
```diff
@@ -642,6 +838,30 @@ class TReplicationService: public TActorBootstrapped<TReplicationService> {
+        } else {
+            StatsWakeupScheduled = false;
+        }
+        if (!SessionsToWake.empty() && !StatsWakeupScheduled) {
```
```suggestion

        if (!SessionsToWake.empty() && !StatsWakeupScheduled) {
```

CyberROFL commented on ydb/core/tx/replication/controller/target_with_stream.cpp:
```diff
@@ -113,11 +123,103 @@ class TWorkerRegistar: public TActorBootstrapped<TWorkerRegistar> {
+        default:
+            return false;
+    };
+    return true;
```
```suggestion
    }

    return true;
```

CyberROFL commented on ydb/core/tx/replication/controller/target_with_stream.cpp:
```diff
@@ -11,10 +11,15 @@
 #include <ydb/library/actors/core/hfunc.h>
 
+#include <ydb/core/protos/metrics_config.pb.h>
+#include <ydb/public/api/protos/draft/ydb_replication.pb.h>
```
Непонятно, для чего тут эта зависимость.

CyberROFL commented on ydb/core/tx/replication/controller/target_transfer.h:
```diff
@@ -2,37 +2,52 @@
 
+#include <ydb/core/base/appdata.h>
+#include <ydb/core/base/counters.h>
+#include <ydb/core/tx/replication/service/worker.h>
```
Нет. Выдели общий код и положи в common.

FloatingCrowbar replied:
Речь про конкретно worker.h - ? Убрал

CyberROFL commented on ydb/core/tx/replication/controller/target_transfer.cpp:
```diff
@@ -1,15 +1,222 @@
+};
+
+struct TTransferCounters: public TTragetWithStreamCounters {
+
```
```suggestion
```

FloatingCrowbar committed:
Apply suggestions from code review

Co-authored-by: Ilnaz Nizametdinov <i.nizametdinov@gmail.com>

FloatingCrowbar committed:
Improve on review

FloatingCrowbar committed:
Refactory after review

FloatingCrowbar committed:
Merge branch 'main' into transfer-obss

FloatingCrowbar committed:
Fix the bug

FloatingCrowbar committed:
Merge branch 'main' into transfer-obss

FloatingCrowbar committed:
cleanup

FloatingCrowbar committed:
Fix normalization in viewer test

pnv1 reviewed (approved):
C++ SDK part LGTM
